src/mqtt.rs

```diff
@@ -13,14 +15,221 @@ pub struct Mqtt {
+        self.inverter_state.iter().find(
+            |inv| inv.port_id == pv_port
+        ).unwrap_or(&InverterState::default()).clone()
+    }
+
+    fn to_json_payload(&self) -> serde_json::Value {
+        // when modifying this function, modify the sensor config in create_device_config accordingly
+        let mut json = json!({
+            "dtu_sn": self.dtu_sn.iter().map(|v| v.to_string()).collect::<String>(),
+            "pv_current_power": format!("{:.2}", self.pv_current_power as f32 * 0.1),
+            "pv_daily_yield": self.pv_daily_yield,
+            "efficiency": self.get_total_efficiency(),
+        });
+
+        // Convert each PortState to json
+        for port in self.port_state.iter() {
+            json[format!("pv_{}_vol", port.pv_port)] = format!("{:.2}", port.pv_vol as f32 * 0.1).into();
+            json[format!("pv_{}_cur", port.pv_port)] = format!("{:.2}", port.pv_cur as f32 * 0.1).into();
+            json[format!("pv_{}_power", port.pv_port)] = format!("{:.2}", port.pv_power as f32 * 0.1).into();
+            json[format!("pv_{}_energy_total", port.pv_port)] = port.pv_energy_total.into();
+            json[format!("pv_{}_daily_yield", port.pv_port)] = port.pv_daily_yield.into();
+            
+            let inverter = &self.get_inverter(port.pv_port);
+
+            json[format!("inv_{}_grid_voltage", port.pv_port)] = format!("{:.2}", inverter.grid_voltage as f32 * 0.1).into();
+            json[format!("inv_{}_grid_freq", port.pv_port)] = format!("{:.2}", inverter.grid_freq as f32 * 0.1).into();
+            json[format!("inv_{}_pv_current_power", port.pv_port)] = format!("{:.2}", inverter.pv_current_power as f32 * 0.1).into();
+            json[format!("inv_{}_temperature", port.pv_port)] = format!("{:.2}", inverter.temperature as f32 * 0.1).into();
+
+            // Efficiency: Requires data from both PortState and InverterState
+            let efficiency = (port.pv_power as f32) / (inverter.pv_current_power as f32) * 100.0;
+            json[format!("inv_{}_efficiency", port.pv_port)] = format!("{:.2}", efficiency).into();
+        }
+
+        json
+    }
+}
+
+/// `DeviceConfig` is used to define the configuration for a Home Assistant device
+/// in the MQTT discovery protocol and is used to group entities together.
+/// 
+#[derive(Serialize, Clone)]
+pub struct DeviceConfig {
+    name: String,
+    model: String,
+    identifiers: Vec<String>,
+    manufacturer: String,
+    sw_version: String,  // Software version of the application that supplies the discovered MQTT item.
+}
+
+
+impl DeviceConfig {
+    pub fn new(name: String, model: String, identifiers: Vec<String>) -> Self {
+        Self {
+            name,
+            model,
+            identifiers,
+            manufacturer: "Hoymiles".to_string(),
+            /// Rust compiler sets the CARGO_PKG_VERSION environment from the Cargo.toml . 
+            sw_version: env!("CARGO_PKG_VERSION").to_string(),
```

Comment:
This is arguably not part the version of the inverter device

Reply:
No, but I thought it makes sense to log the version of the publisher here (we can also change it in the future in case we figure out a way to extract the sw version of the inverter)